### PR TITLE
Decode EFB copies used as paletted textures.

### PR DESCRIPTION
**Pull request by magumagu**
A number of games make an EFB copy in I4/I8 format, then use it as a
texture in C4/C8 format.  Detect when this happens, and decode the copy on
the GPU using the specified palette.

This has a few advantages: it allows using EFB2Tex for a few more games,
it, it preserves the resolution of scaled EFB copies, and it's probably a
bit faster.

D3D only at the moment, but porting to OpenGL should be straightforward.  I'm not as familiar with the OpenGL API, though, so I would appreciate some help with that.

The changes to the texture cache (TextureCacheBase.h/cpp) are the part that really needs to be reviewed closely; I'm not really confident this is the right direction.

Please don't merge yet; this will cause regressions on OpenGL as-is.

### CODE REVIEW

**Comment by JMC47** [comment]
Pinging @degasus and @Sonicadvance1 for an OpenGL implementation.

**Comment by delroth** [comment]
https://fifoci.dolphin-emu.org/compare/105059-104832/

Very nice :)

**Comment by delroth** [comment]
@dolphin-emu-bot rebuild

**Comment by Mullin** [comment]
Very nice !

File: Source/Core/VideoCommon/TextureCacheBase.h [hunk 1]
```diff
@@ -43,13 +43,54 @@ class TextureCache
...
+		{
+			// A texture in RAM
+			RAM,
+			// A texture loaded into TMEM
+			TMEM,
+			// A texture in RAM, fully decoded using a palette.
+			RAM_PALETTE,
+			// A texture with two TMEM addresses.
+			// May be either decoded with a palette or an RGBA8 texture with two parts.
+			DUAL_TMEM,
```
Comment (phire) [inline Source/Core/VideoCommon/TextureCacheBase.h]: I'm not sure it's wise having two meanings for this.

File: Source/Core/VideoBackends/D3D/TextureCache.cpp [hunk 2]
```diff
@@ -179,17 +181,165 @@ void TextureCache::TCacheEntry::FromRenderTarget(u32 dstAddr, unsigned int dstFo
...
+	}
+	return float4(r, g, b, a) / 255;
+}
+
+float4 DecodePixel_RGB565(uint val)
+{
+	int r, g, b, a;
+	r = Convert5To8((val >> 11) & 0x1f);
+	g = Convert6To8((val >> 5) & 0x3f);
+	b = Convert5To8((val)& 0x1f);
```
Comment (phire) [inline Source/Core/VideoBackends/D3D/TextureCache.cpp]: Space between ) and &

File: Source/Core/VideoBackends/D3D/TextureCache.cpp [hunk 3]
```diff
@@ -179,17 +181,165 @@ void TextureCache::TCacheEntry::FromRenderTarget(u32 dstAddr, unsigned int dstFo
...
+R"HLSL(
+sampler samp0 : register(s0);
+Texture2DArray Tex0 : register(t0);
+Buffer<uint> Tex1 : register(t1);
+uniform float Multiply;
+
+uint Convert3To8(uint v)
+{
+	// Swizzle bits: 00000123 -> 12312312
+	return (v << 5) | (v << 2) | (v >> 1);
```
Comment (phire) [inline Source/Core/VideoBackends/D3D/TextureCache.cpp]: There is nowhere you don't use these functions without an &7 (or however many bits)

Might as well build the masking into these functions.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 4]
```diff
@@ -324,61 +315,105 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
...
+				textures[paletteDecodedID] = decoded_entry;
+				entry = decoded_entry;
+			}
 			return ReturnEntry(stage, entry);
 		}
 
-		// 2. b) For normal textures, all texture parameters need to match
-		if (address == entry->addr && tex_hash == entry->hash && full_format == entry->format &&
+		// For normal textures, all texture parameters need to match.
+		if ((tex_hash ^ tlut_hash) == entry->hash && full_format == entry->format &&
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: I think we should build up a generic format converison API. So if we remove the "full_format == entry->format" check here, we're able to call a generic format convertion function.

Does the bits per pixel usually match? If so, then we don't have to care about different scaling issues with tiling...

**Comment by degasus** [comment]
Was there a reason to sqaush this implementation with your deferred freeing of textures? imo for the deferred freeing, it's better to just lock this texture for at least a draw call, or just until the next frame, or better don't remove them from the texture cache for at least a frame ( #2001 )

File: Source/Core/VideoBackends/OGL/TextureCache.cpp [hunk 5]
```diff
@@ -359,4 +359,10 @@ void TextureCache::DeleteShaders()
 	s_DepthMatrixProgram.Destroy();
 }
 
+void TextureCache::ConvertTexture(TextureCache::TCacheEntryBase* entry, TCacheEntryBase* unconverted, void* palette, TlutFormat format)
+{
+	// TODO: Implement.
+	return;
```
Comment (degasus) [inline Source/Core/VideoBackends/OGL/TextureCache.cpp]: It's fine to only support D3D, but please change this to "return false;"

File: Source/Core/VideoBackends/D3D/TextureCache.cpp [hunk 6]
```diff
@@ -179,17 +181,165 @@ void TextureCache::TCacheEntry::FromRenderTarget(u32 dstAddr, unsigned int dstFo
...
 	g_encoder = new PSTextureEncoder;
 	g_encoder->Init();
+
+	palette_buf = nullptr;
+	palette_buf_srv = nullptr;
+	palette_uniform = nullptr;
+	palette_pixel_shader[GX_TL_IA8] = GetConvertShader("IA8");
+	palette_pixel_shader[GX_TL_RGB565] = GetConvertShader("RGB565");
+	palette_pixel_shader[GX_TL_RGB5A3] = GetConvertShader("RGB5A3");
+	auto lutBd = CD3D11_BUFFER_DESC(sizeof(u16) * 256, D3D11_BIND_SHADER_RESOURCE);
```
Comment (degasus) [inline Source/Core/VideoBackends/D3D/TextureCache.cpp]: What will happen for I14X2?

**Comment by Linktothepast** [comment]
Apart from the twilight princess map that is upscaled and working with efb to texture, seems that rogue squadron and rebel strike targeting computer works perfectly fine too with efb to texture, unfortunately those games still need efb to ram for a few effects like distant fog etc. They are quite playable with efb to texture bearing a few such glitches.

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 4 above)
Comment (magumagu) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Bits per pixel matches for all the EFB copy cases I've seen.

We might be able to build up a "generic" format conversion?  I'll have to think about that a bit more.  Not sure it makes sense to use the same function for EFB copies and other textures.

Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: It would be still faster than to reupload the texture. But I also don't see the big usage. Maybe just upload all paletted as I8 first and convert them to C8 afterwards ;) So the I8 can be cached

File: Source/Core/VideoBackends/D3D/TextureCache.cpp (hunk 6 above)
Comment (magumagu) [inline Source/Core/VideoBackends/D3D/TextureCache.cpp]: It won't work.  I haven't seen anything try to use it.   Making it work would probably require changing the way we generate EFB copies.

**Comment by magumagu** [comment]
Updated to get rid of the deferred freeing; some quick testing shows it isn't actually necessary.  (Hopefully, this doesn't break anything.)

**Comment by JMC47** [comment]
Metroid Prime 1 won't need EFB2RAM at all.

Metroid Prime 1 - 3 will upscale the other visors with this.

**Comment by mimimi085181** [comment]
Okami is another game that won't need efb2ram anymore after this. Not even the drawing part needs efb2ram.

File: Externals/libusb/msvc/libusb_static_2013.vcxproj [hunk 7]
```diff
@@ -137,7 +137,6 @@
     <ClCompile Include="..\libusb\hotplug.c" />
     <ClCompile Include="..\libusb\io.c" />
     <ClCompile Include="..\libusb\os\poll_windows.c" />
-    <ClCompile Include="..\libusb\strerror.c" />
```
Comment (delroth) [inline Externals/libusb/msvc/libusb_static_2013.vcxproj]: That looks unrelated? Care to split these changes out?

File: Source/Core/VideoBackends/D3D/TextureCache.cpp (hunk 6 above)
Comment (delroth) [inline Source/Core/VideoBackends/D3D/TextureCache.cpp]: Add a TODO in the code explaining this.

File: Source/Core/VideoBackends/D3D/TextureCache.h [hunk 8]
```diff
@@ -42,8 +42,15 @@ class TextureCache : public ::TextureCache
 
 	u64 EncodeToRamFromTexture(u32 address, void* source_texture, u32 SourceW, u32 SourceH, bool bFromZBuffer, bool bIsIntensityFmt, u32 copyfmt, int bScaleByHalf, const EFBRectangle& source) {return 0;};
 
+	virtual void ConvertTexture(TCacheEntryBase* entry, TCacheEntryBase* unconverted, void* palette, TlutFormat format) override;
```
Comment (delroth) [inline Source/Core/VideoBackends/D3D/TextureCache.h]: Drop the "virtual" since "override" is already there.

File: Source/Core/VideoBackends/D3D/TextureCache.h [hunk 9]
```diff
@@ -42,8 +42,15 @@ class TextureCache : public ::TextureCache
...
+
 	void CompileShaders() override { }
 	void DeleteShaders() override { }
+
+	ID3D11Buffer *palette_buf;
```
Comment (delroth) [inline Source/Core/VideoBackends/D3D/TextureCache.h]: The style in this file is pointer star next to type. Please fix.

File: Source/Core/VideoBackends/OGL/TextureCache.cpp (hunk 5 above)
Comment (delroth) [inline Source/Core/VideoBackends/OGL/TextureCache.cpp]: Or get someone to implement it in GL :)

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 10]
```diff
@@ -303,13 +286,16 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
...
 
-	u32 texID = address;
 	// Hash assigned to texcache entry (also used to generate filenames used for texture dumping and custom texture lookup)
 	u64 tex_hash = TEXHASH_INVALID;
 
 	u32 full_format = texformat;
 
 	const bool isPaletteTexture = (texformat == GX_TF_C4 || texformat == GX_TF_C8 || texformat == GX_TF_C14X2);
+
+	if (isPaletteTexture && tlutfmt > GX_TL_RGB5A3)
```
Comment (delroth) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Again, add a TODO to explain why we don't handle these other formats.

File: Source/Core/VideoCommon/VertexManagerBase.cpp [hunk 11]
```diff
@@ -209,6 +209,7 @@ void VertexManager::Flush()
 				if (bpmem.tevind[i].IsActive() && bpmem.tevind[i].bt < bpmem.genMode.numindstages)
 					usedtextures[bpmem.tevindref.getTexMap(bpmem.tevind[i].bt)] = true;
 
+		TextureCache::UnbindTextures();
```
Comment (delroth) [inline Source/Core/VideoCommon/VertexManagerBase.cpp]: How is the overhead of this unbind/bind for each flush? @degasus

File: Source/Core/VideoBackends/D3D/TextureCache.cpp [hunk 12]
```diff
@@ -179,17 +181,165 @@ void TextureCache::TCacheEntry::FromRenderTarget(u32 dstAddr, unsigned int dstFo
...
+		a=0xFF;
+	}
+	else
+	{
+		a=Convert3To8((val>>12) & 0x7);
+		r=Convert4To8((val>>8 ) & 0xf);
+		g=Convert4To8((val>>4 ) & 0xf);
+		b=Convert4To8((val    ) & 0xf);
+	}
+	return float4(r, g, b, a) / 255;
```
Comment (FioraAeterna) [inline Source/Core/VideoBackends/D3D/TextureCache.cpp]: If we're converting to float, should this be 255.f?  (pedantry)

File: Externals/libusb/msvc/libusb_static_2013.vcxproj (hunk 7 above)
Comment (magumagu) [inline Externals/libusb/msvc/libusb_static_2013.vcxproj]: This is PR #2063.

File: Source/Core/VideoCommon/VertexManagerBase.cpp (hunk 11 above)
Comment (magumagu) [inline Source/Core/VideoCommon/VertexManagerBase.cpp]: This is basically zero additional overhead; I just needed to reorder the API calls so that I can bind textures in TextureCache::Load without worrying about messing up existing bindings.

Comment (magumagu) [inline Source/Core/VideoCommon/VertexManagerBase.cpp]: (A potential optimization would be to skip recomputing a binding if it's "the same" as it was the last time we flushed, but we don't do that with or without this PR.)

**Comment by Linktothepast** [comment]
JMC47 Metroid Prime still needs efb to ram, there is texture corruption without it in the Chozo ruins that i tested regardless of scanner working .

**Comment by JMC47** [comment]
Is it worth the slowdown at that point to have EFB2RAM on?

**Comment by Linktothepast** [comment]
http://prntscr.com/66uv94 See for yourself and judge, i get something like that all over the place.

**Comment by Tinob** [comment]
@Linktothepast Have you tested increasing cache samples to see if that fixes the issue?

**Comment by Linktothepast** [comment]
Nope, it happens in the safest setting too but seems to be fixed temporarily when moving the slider.

**Comment by delroth** [comment]
@Linktothepast do you happen to have a fifolog of that?

**Comment by JMC47** [comment]
I haven't run into those bugs in my quick testing of EFB2Tex in Metroid Prime... I guess I'll go through the Chozo Ruins and see what happens.

**Comment by Linktothepast** [comment]
https://www.mediafire.com/?idb9wwef9n1l99m there you go, a five frame fifo, it happens with efb to texture and d3d that i tested (didn't check opengl).

**Comment by mimimi085181** [comment]
@Linktothepast: Can you test if this happens the master build as well? It might be unrelated to this PR. If it works there, 1% chance a recent change fixed it, can you test the master version the PR is based on then?

**Comment by Linktothepast** [comment]
Same thing in master unfortunately.

**Comment by delroth** [comment]
Oh, so that shouldn't be considered a blocker for this PR. Cool.

Can someone maybe port it to GL? :) Or at least fix the GL related bugs that it introduces.

File: Externals/libusb/msvc/libusb_static_2013.vcxproj (hunk 7 above)
Comment (delroth) [inline Externals/libusb/msvc/libusb_static_2013.vcxproj]: #2063 has now been merged. Please remove these changes from this PR.

**Comment by Linktothepast** [comment]
Eh, no i just told jmc that efb to ram is still needed for the game due to the above issue.

File: Source/Core/VideoBackends/D3D/TextureCache.cpp (hunk 3 above)
Comment (magumagu) [inline Source/Core/VideoBackends/D3D/TextureCache.cpp]: I would prefer to keep the extract operation (right shift, then mask) together; I guess I could make an Extract3To8 helper.

**Comment by magumagu** [comment]
Added fallback path for OpenGL; should be unchanged from master.

**Comment by magumagu** [comment]
This could probably use one more round of testing to make sure I didn't break anything merging with PR #2001.

Also, JMC47 discovered an issue with the RS2 pause screen with EFB2Tex; I'm not sure what's causing that.

**Comment by magumagu** [comment]
Figured out what's going on with RS2: it has to do with the interaction between BPMEM_PRELOAD_MODE and EFB copies.  The way we implement BPMEM_PRELOAD_MODE is a memcpy from main memory to texMem; however, we don't notify the texcache that this is happening, so it loses track of the relationship.  Or rather, it *should* lose track of the relationship, but on master we use the specified address in main memory as a hint to find the right entry even though it doesn't actually have any meaning (?).

Anyway, I'm going to have to mess with the texture cache side of this a bit more before merging this.

**Comment by mimimi085181** [comment]
As far as i can tell, the rebase on top of #2001 went well. Tested Castlevania 3, Metroid Prime 3(fonts), New Super Mario Bros, Okami and RS2, all ran as expected of #2001 and this PR.

**Comment by magumagu** [comment]
Rewrote the texture cache stuff; should fix the RS2 regression.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 13]
```diff
@@ -406,6 +402,23 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
 		++iter;
 	}
 
+	if (unconverted_copy != textures.end())
+	{
+		// Perform palette decoding.
+		TCacheEntryBase *entry = unconverted_copy->second;
+		TCacheEntryBase *decoded_entry = AllocateTexture(entry->config);
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Maybe we should add a check if the config only allocate one mipmap layer. It's always true right now, but very fragile.

Comment (magumagu) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Directly copying the configuration of the input is fragile, but adding a check seems sort of awkward. Changed to explicitly set up the configuration.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 14]
```diff
@@ -406,6 +402,29 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
...
+	if (unconverted_copy != textures.end())
+	{
+		// Perform palette decoding.
+		TCacheEntryBase *entry = unconverted_copy->second;
+
+		TCacheEntryConfig config;
+		config.rendertarget = true;
+		config.width = entry->config.width;
+		config.height = entry->config.height;
+		config.layers = FramebufferManagerBase::GetEFBLayers();
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Uh, I did miss this: Does this PR works fine with stereo? @Armada651

**Comment by DarthVitrial** [comment]
So, does this PR fix it with OpenGL too, or is that coming in the future?
(Sorry if I'm not supposed to post here)
